### PR TITLE
Windows File System Unicode Fixes and correctly expand home directory in ATTACH/DBInstanceCache

### LINKED ISSUE #5499: Windows Python connect method is requiring drive name in .6. Was working without drive name in .5.9
### What happens?

On windows when connecting via Python in the connect method I have to pass in the drive name in .60. 
In .5.9 drivename was not nessecary. 
E.g.. if dbfile is here f:\temp\sample.db  
I can't pass in \\temp\\sample.db or /temp/sample.db.  It prepends current path. 
e.g. if my current path is f:/myprogram it tries to open f:/myprogram/temp/sample.db  and fails.
This behavior is different from .5.9 

### To Reproduce

Create a db in a directory (f:\temp\sample.db). I acutally didn't check C drive but in my F drive.

IN a diffrent directory e.g.
f:/myprogram create a python program that just does a connect and passes in  \\temp\\sample.db or /temp/sample.db.
It will fail with cannot open in .6 but will work in .5.9

Now do a connect with
f:\\temp\\sample.db or f:/temp/sample.db.
It will work with .6

### OS:

windows 10 64

### DuckDB Version:

.6

### DuckDB Client:

Python

### Full Name:

Bhala Ghatate

### Affiliation:

Daman

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### LINKED ISSUE #6573: accent in file path causing IOException
### What happens?

duckdb.connect() causes an IOException when the full filepath contains an "é" character, but a relative path is provided. If the section of the path containing the "é" character is provided, it works fine.

### To Reproduce

image

### OS:

Windows10 x64

### DuckDB Version:

0.7.1, duckdb-0.7.2.dev225

### DuckDB Client:

Python

### Full Name:

Clement Sreeves

### Affiliation:

Personal Project

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### LINKED ISSUE #6931: [Python] Empty IOException upon installing JSON extension
### What happens?

Upon installation of JSON extension in DuckDB an empty `IOException` with neither stack trace nor error message gets thrown, making it difficult to debug it further. Here is a simplest example of the failing code:

```python
import duckdb

if __name__ == "__main__":
    # sys.setdlopenflags(os.RTLD_GLOBAL | os.RTLD_NOW)  # Running on Windows
    con = duckdb.connect()
    con.execute("INSTALL 'json';")
```

The process also fails when attempting installation from file:

```python
import duckdb

if __name__ == "__main__":
    # File lies in CWD
    duckdb.install_extension("json.duckdb_extension")
```
In both cases, all the error information returned is:
```
Traceback (most recent call last):
  File "c:\dev\projects\openai\debug.py", line 10, in <module>
    con.execute("INSTALL 'json';")
duckdb.IOException
```
It does not look to me like a problem finding the installation file not a syntax error in my SQL query. Otherwise the DB seems to function properly; I am able to connect, create a test table, insert data and fetch it via the `fetchdf` method. I am using release `0.7.1` but I got the same error with the `0.7.0`, `0.6.1` and `0.7.2.dev1256` versions.

### To Reproduce

1) Create a `pipenv` environment from the following `Pipenv` file:

````toml
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
duckdb = "*"

[dev-packages]

[requires]
python_version = "3.10"
````
2) run the following code snippet:

```python
import duckdb

if __name__ == "__main__":
    con = duckdb.connect()
    con.execute("INSTALL 'json';")
```

### OS:

Windows 10

### DuckDB Version:

0.7.1

### DuckDB Client:

Python

### Full Name:

Daniel Moseguí i González

### Affiliation:

paiqo GmbH

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by Mytherin**
This PR is a continuation of #5852

Fixes #5499
Fixes #6573
Fixes #6931
Supersedes #7266

###### Windows Unicode Fixes
This PR fixes unicode handling on Windows for various system calls by correctly using the `W` variant were necessary. These places are:

* `getenv` -> `_wgetenv`
* `SetCurrentDirectory` -> `SetCurrentDirectoryW`
* `GetCurrentDirectory` -> `GetCurrentDirectoryW`

This allows us to correctly get/set the working directory, as well as get the home directory of the user, even if they contain unicode characters.

In addition, for extension installation we use our own file system APIs instead of `iostream`, as [`iostream` is not unicode aware either on Windows](https://stackoverflow.com/questions/821873/how-to-open-an-stdfstream-ofstream-or-ifstream-with-a-unicode-filename).

This fixes #6573 and #6931.

###### Correctly Expand Home Directory & Normalize Absolute Paths

Similar to how we allow `~` to be used as the home directory when reading from files, we now correctly expand the home directory when opening databases (either through attach, the storage manager, or through the `DBInstanceCache`). This implements #7266 for all clients instead of only the JDBC client.

For #5499 we fix the normalization of absolute paths in the `DBInstanceCache` which caused this regression in the first place.

### CODE REVIEW

File: src/include/duckdb/common/file_system.hpp [hunk 1]
```diff
@@ -187,6 +190,9 @@ class FileSystem {
 	//! Extract the name of a file (e.g if the input is lib/example.dll the name is 'example.dll')
 	DUCKDB_API static string ExtractName(const string &path);
 
+	//! Returns the value of an environment variable - or the empty string if it is not set
+	DUCKDB_API static string GetEnvVariable(const string &name);
```
Comment (Mause) [inline src/include/duckdb/common/file_system.hpp:194]: This should probably not be filesystem specific, and be used here too: https://github.com/duckdb/duckdb/blob/e466d90508a4422179068bb3a94c0c44f239155e/extension/httpfs/s3fs.cpp#L145

**Review by Mause — CHANGES_REQUESTED** [review]

File: src/include/duckdb/common/file_system.hpp (hunk 1 above)
Comment (Mytherin) [reply src/include/duckdb/common/file_system.hpp:194]: We should clean this up in the future - but currently the `FileSystem` class holds almost all OS-specific code. In this case as a static method it is not file-system specific but only OS-specific.

I will leave this as-is for now in light of trying to get this finished before the release, but that definitely should also be updated later on.

**Comment by Tishj** [comment]
I wonder if this also fixes https://github.com/duckdb/duckdb/issues/7246

**Comment by Mytherin** [comment]
I think that's unlikely as that is a data-issue, not a file system issue.

**Pull request merged**